src/twisted/conch/insults/window.py

```diff
@@ -109,7 +109,7 @@ def functionKeyReceived(self, keyID, modifier):
         name = keyID
         if not isinstance(keyID, str):
             name = name.decode("utf-8")
-        func = getattr(self, "func_" + name, None)
+        func = getattr(self, "func_" + name[1:-1], None)
```

Comment:
So... this was introduced in https://github.com/twisted/twisted/pull/648

Before that PR, we had `func = getattr(self, 'func_' + keyID.name, None)`

--------

Why remove the `F` prefix?

This looks like breaking backward compatibility.

Previously the functions that were dispached were named `func_f12` ... with this change the need to be named `func_12` 

-----

I might be missing something here.

Reply:
It's not removing "prefix".  Before 6659c6b4f0290b83d5978e93ed8aff3d4878f649 `FUNCTION_KEYS` were defined as:
```python
_KEY_NAMES = ('UP_ARROW', 'DOWN_ARROW', 'RIGHT_ARROW', 'LEFT_ARROW',
              'HOME', 'INSERT', 'DELETE', 'END', 'PGUP', 'PGDN', 'NUMPAD_MIDDLE',
              'F1', 'F2', 'F3', 'F4', 'F5', 'F6', 'F7', 'F8', 'F9',
              'F10', 'F11', 'F12',
              'ALT', 'SHIFT', 'CONTROL')

FUNCTION_KEYS = [_const(_name) for _name in _KEY_NAMES]
```
these `const` instances were then assigned using their names to `insults.ServerProtocol` using
```python
for name, const in zip(_KEY_NAMES, FUNCTION_KEYS):
    setattr(ServerProtocol, name, const)
```

and `const` instance was passed down to `keystrokeReceived` when appropriate key was received, so (before 5d2ff6938cd1fcfce9e0861ae2c3a562c21b8156)  you could use
```python
func = getattr(self, "func_" + keyID.name, None)
```
and you ended up with proper function reference.

However AFTER 6659c6b4f0290b83d5978e93ed8aff3d4878f649 `FUNCTION_KEYS` are defined as
```python
FUNCTION_KEYS = [_const(_name).__bytes__() for _name in _KEY_NAMES]
```
and this means you have something like:
```python
[b'[UP_ARROW]', b'[DOWN_ARROW]', b'[RIGHT_ARROW]', b'[LEFT_ARROW]', b'[HOME]', ...]
```
this end up in `ServerProtocol`:
```python
>>> ServerProtocol.DOWN_ARROW
b'[DOWN_ARROW]'
```
and is passed down to `keystrokeReceived` (added prints to `ButtonDemo.keystrokeReceived` in `window.tac` to show this)
```
2023-12-01T13:51:55+0100 [stdout#info] keystrokeReceived: b'f' None
2023-12-01T13:51:57+0100 [stdout#info] keystrokeReceived: b'\t' None
2023-12-01T13:51:58+0100 [stdout#info] keystrokeReceived: b'\t' None
2023-12-01T13:51:59+0100 [stdout#info] keystrokeReceived: b'[DOWN_ARROW]' None
2023-12-01T13:52:00+0100 [stdout#info] keystrokeReceived: b'[UP_ARROW]' None
```

This all means now (after 5d2ff6938cd1fcfce9e0861ae2c3a562c21b8156) when `functionKeyReceived` is called with e.g. `b'[DOWN_ARROW]'`...
```python
def functionKeyReceived(self, keyID, modifier):
    name = keyID
    if not isinstance(keyID, str):
        name = name.decode("utf-8")
    func = getattr(self, "func_" + name, None)
    if func is not None:
        func(modifier)
```
... it won't (obviously) find any `func_[DOWN_ARROW]` method.

TL;DR: if something broke backwards compatibility, it was patch 6659c6b4f0290b83d5978e93ed8aff3d4878f649, as it changed the way `FUNCTION_KEYS` are defined, followed by 5d2ff6938cd1fcfce9e0861ae2c3a562c21b8156, where it was not noticed `keyID` is now in form of `b'[<whatever>]'`. My patch simply strips those brackets to get valid identifier.

Reply:
Anyway this all 
```python
class _const:
    """
    @ivar name: A string naming this constant
    """

    def __init__(self, name: str) -> None:
        self.name = name

    def __repr__(self) -> str:
        return "[" + self.name + "]"

    def __bytes__(self) -> bytes:
        return ("[" + self.name + "]").encode("ascii")


FUNCTION_KEYS = [_const(_name).__bytes__() for _name in _KEY_NAMES]
```
in [twisted.conch.insults.insults.py#L435](https://github.com/twisted/twisted/blob/311d8dbaf40ddeb887cd3edab1333a3180e9d4d9/src/twisted/conch/insults/insults.py#L435) is just superfluous, as the only use of it now is to convert `"THIS"` to `b"[THIS]"` and could be replaced with:

```python
FUNCTION_KEYS = [f"[{x}]".encode("ascii") for x in _KEY_NAMES]
```
I'll extend this PR to include this change.